### ISSUE TITLE
d2i_CV_RESPONSE fails for CMS encapsulated SCVP-CV responses

### ISSUE THREAD

**Issue by kartikkrishnan11** [openssl/openssl#22370]
The following code fails at d2i_CV_RESPONSE due to openssl raising error for implicit tagging.

const unsigned char *pData = NULL;
ASN1_OCTET_STRING **pOctetStr = NULL;
pData = (const unsigned char*) buffer;
CMS_ContentInfo * cmsResponseContentInfo = d2i_CMS_ContentInfo(NULL, &pData, length);
if (!cmsResponseContentInfo) return retval;

if (CMS_get0_eContentType(cmsResponseContentInfo) != OBJ_nid2obj(NID_id_ct_scvp_certValResponse)) {
	CMS_ContentInfo_free(pCmsResponse->cmsResponseContentInfo);
	return retval;
}
pOctetStr = CMS_get0_content(cmsResponseContentInfo);
if (pOctetStr == NULL || *pOctetStr ==NULL) return retval;

pData = (*pOctetStr)->data;
//The following line of code fails at the d2i and returns NULL.
CV_RESPONSE * cvResponse = d2i_CV_RESPONSE(NULL, &pData, (*pOctetStr)->length);

The cause of this failure was identified to be the following lines of code in crypto/asn1_tasn_dec.c
In asn1_item_embed_d2i function, for 

 case ASN1_ITYPE_CHOICE:
        /*
         * It never makes sense for CHOICE types to have implicit tagging, so
         * if tag != -1, then this looks like an error in the template.
         */
        if (tag != -1) {
            ERR_raise(ERR_LIB_ASN1, ASN1_R_BAD_TEMPLATE);
            goto err;
        }

This worked in OpenSSL 1.0.2 and the check for tag != -1 wasn't present there. In our case, we have many structures within the SCVP CV RESPONSE which have implicit tagging. Example:
ASN1_CHOICE(CERT_REFERENCE) = {
    ASN1_IMP(CERT_REFERENCE, d.pkc, PKC_REFERENCE, 0),
    ASN1_IMP(CERT_REFERENCE, d.ac, AC_REFERENCE, 1)
} ASN1_CHOICE_END(CERT_REFERENCE)

ASN1_CHOICE(PKC_REFERENCE) = {
    ASN1_IMP(PKC_REFERENCE, d.cert, X509, 0),
    ASN1_IMP(PKC_REFERENCE, d.pkcRef, SCVP_CERT_ID, 1)
} ASN1_CHOICE_END(PKC_REFERENCE)

ASN1_CHOICE(AC_REFERENCE) = {
    ASN1_IMP(AC_REFERENCE, d.attrCert, ATTR_CERT, 2),
    ASN1_IMP(AC_REFERENCE, d.acRef, SCVP_CERT_ID, 3)
} ASN1_CHOICE_END(AC_REFERENCE)

The converse is also true in the i2d_CV_RESPONSE failing due to the tag check crypto/asn1/tasn_enc.c
In ASN1_item_ex_i2d function for the case ASN1_TYPE_CHOICE

**Comment by t8m** [comment]
Which one of the choices is chosen if the tagging is implicit? How should the decoder know?

**Comment by kartikkrishnan11** [comment]
The CERT_REFERENCE structure itself is defined as follows:
typedef struct CERT_REFERENCE_st {
    int type;
    union {
        PKC_REFERENCE *pkc;    /* tag == 0 */
        AC_REFERENCE *ac;    /* tag == 1 */
    } d;
} CERT_REFERENCE;
So, the type holds which type the CERT_REFERENCE contains and we use it accordingly.

**Comment by totszwai** [comment]
Does anyone know a workaround or is there a workaround for this?

**Comment by totszwai** [comment]
I got my code to work (archiving the same successful behavior back in the days) by commenting out all the `if (tag != -1)` from both the encoder and decoder of the tasn.

![image](https://github.com/user-attachments/assets/5a66bbad-6018-48c7-846f-75b3c757da80)

Essentially, this revert back to the old behavior around version `1.1.0f ` (more or less around the time that the `tag != -1` was added?)

What was the reason the check was introduced and what benefit did it bring? Am I actually  doing the right thing?

**Comment by t8m** [comment]
I do not think simply removing the checks would be the right patch. These checks were added for purpose and are related to CVE-2020-1971.

See the commit messages:

https://github.com/openssl/openssl/commit/41d62636fd996c031c0c7cef746476278583dc9e
https://github.com/openssl/openssl/commit/1ecc76f6746cefd502c7e9000bdfa4e5d7911386
https://github.com/openssl/openssl/commit/f960d81215ebf3f65e03d4d5d857fb9b666d6920
https://github.com/openssl/openssl/commit/aa0ad2011d3e7ad8a611da274ef7d9c7706e289b

**Comment by t8m** [comment]
Hmm... looking at RFC 5055, should the definition of CertReference be instead:
```
ASN1_CHOICE(CERT_REFERENCE) = {
ASN1_SIMPLE(CERT_REFERENCE, d.pkc, PKC_REFERENCE),
ASN1_SIMPLE(CERT_REFERENCE, d.ac, AC_REFERENCE)
} ASN1_CHOICE_END(CERT_REFERENCE)
```

**Comment by totszwai** [comment]
> I do not think simply removing the checks would be the right patch. These checks were added for purpose and are related to [CVE-2020-1971](https://github.com/advisories/GHSA-whf2-mq76-2fhv).
> 
> See the commit messages:
> 
> [41d6263](https://github.com/openssl/openssl/commit/41d62636fd996c031c0c7cef746476278583dc9e) [1ecc76f](https://github.com/openssl/openssl/commit/1ecc76f6746cefd502c7e9000bdfa4e5d7911386) [f960d81](https://github.com/openssl/openssl/commit/f960d81215ebf3f65e03d4d5d857fb9b666d6920) [aa0ad20](https://github.com/openssl/openssl/commit/aa0ad2011d3e7ad8a611da274ef7d9c7706e289b)

Well, removing those checks (shown in my reply) allowed me to parse an SCVP response generated back in 2016 (which used either the `1.0.2` or `1.1.0f` that was the on. Otherwise the `tag != -1` (which would always be `true` because it is being passed by the [upstream](https://github.com/openssl/openssl/blob/OpenSSL_1_1_1w/crypto/asn1/tasn_dec.c#L643)) will always prevent the flow from continuing like the old way.

**Comment by t8m** [comment]
That unfortunately does not mean that removing those checks is the correct fix. I do not think we are going to revert these commits.
It also can mean that the generated SCVP responses are not matching the standard.

**Comment by totszwai** [comment]
Hum.. quick question @t8m, what is the correct (or in general, just how) to register an explicit TAG with the ASN1? If I am trying to decode and parse the ASN1 structure, is there a way to either:
- Tell `openssl` to bypass the tag check because we instructed it to do so or
- Manipulate the CMS structure before passing it over to `openssl` such that the tag would be a different value other than `-1`?

**Comment by t8m** [comment]
I do not think there is a simple way to do either. I also believe the encoding you've produced with the above definition with old openssl releases is incorrect.

**Comment by totszwai** [comment]
> I do not think there is a simple way to do either. I also believe the encoding you've produced with the above definition with old openssl releases is incorrect.

Yeah, it was incorrect, but we still need to parse the old response; while the old version of openssl before those commits (`tag != -1`) went in, we were still able to parse it just fine but now lost the ability to do so without any means to configure/accept it.

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity leading to implementing or re-prioritizing it for more than 2 months.
